textattack/augmentation/augmenter.py

```diff
@@ -11,40 +15,100 @@ class Augmenter:
-        `self.transformation`.
+        """ Returns all possible augmentations of `text` according to 
+            `self.transformation`.
         """
         tokenized_text = TokenizedText(text, DummyTokenizer())
-        # Get potential transformations for text.
-        transformations = self.transformation(tokenized_text)
-        # Filter out transformations that don't match the constraints.
-        transformations = self._filter_transformations(tokenized_text, transformations)
-        return [t.clean_text() for t in transformations]
+        all_transformations = set()
+        for _ in range(self.transformations_per_example):
+            indices_to_modify = set(range(len(tokenized_text.words)))
+            next_tokenized_text = tokenized_text
+            for __ in range(self.words_to_swap):
+                transformations = []
+                for constraint in self.pre_transformation_constraints:
+                    indices_to_modify = list(set(indices_to_modify) & constraint(tokenized_text, self))
+                while not len(transformations):
+                    # Loop until we find a valid transformation.
+                    if not len(indices_to_modify):
+                        # This occurs when we couldn't find valid transformations
+                        # – either the constraints were too strict, or the number
+                        # of words to swap was too high, or we were just plain 
+                        # unlucky. In any event, don't throw an error, and just 
+                        # don't return anything.
+                        break
+                    replacement_index = random.choice(indices_to_modify)
+                    indices_to_modify.remove(replacement_index)
+                    transformations = self.transformation(next_tokenized_text, indices_to_modify=[replacement_index])
+                    # Get rid of transformations we already have
+                    transformations = [t for t in transformations if t not in all_transformations]
+                    # Filter out transformations that don't match the constraints.
+                    transformations = self._filter_transformations(tokenized_text, transformations)
+                if len(transformations):
+                    next_tokenized_text = random.choice(transformations)
+            all_transformations.add(next_tokenized_text)
+        return [t.clean_text() for t in all_transformations]
     
     def augment_many(self, text_list):
         """
         Returns all possible augmentations of a list of strings according to
         `self.transformation`.
+    
+        Args:
+            text_list (list(string)): a list of strings for data augmentation
+            
+        Returns a list(string) of augmented texts.
         """
         return [self.augment(text) for text in text_list]
         
+    def augment_text_and_ids(self, text_list, id_list, show_progress=True):
```

Comment:
maybe change to augment_text_with_ids or something? this makes it sound like the IDs are being augmented